libs/platform/unit/platform_unit_test.cpp

```diff
@@ -703,6 +707,54 @@ BIT_MASK_TEST(65);
 BIT_MASK_TEST(129);
 BIT_MASK_TEST(1025);
 
+TEST_CASE("completion", "[platform]")
+{
+    _test_helper test_helper;
+
+    auto test = [](bool complete) {
+        struct _completion_context
+        {
+            ebpf_result_t result;
+        } completion_context = {EBPF_PENDING};
+
+        struct _cancellation_context
+        {
+            bool cancelled;
+        } cancellation_context = {false};
+
+        REQUIRE(
+            ebpf_async_set_completion_callback(&completion_context, [](_Inout_ void* context, ebpf_result_t result) {
+                auto completion_context = reinterpret_cast<_completion_context*>(context);
+                completion_context->result = result;
+            }) == EBPF_SUCCESS);
+
+        REQUIRE(ebpf_async_set_cancel_callback(&completion_context, &cancellation_context, [](void* context) {
+                    auto cancellation_context = reinterpret_cast<_cancellation_context*>(context);
+                    cancellation_context->cancelled = true;
+                }) == EBPF_SUCCESS);
+        REQUIRE(completion_context.result == EBPF_PENDING);
+        REQUIRE(!cancellation_context.cancelled);
+
+        if (complete) {
+            REQUIRE(ebpf_async_complete(&completion_context, EBPF_SUCCESS));
+            REQUIRE(completion_context.result == EBPF_SUCCESS);
+            REQUIRE(!cancellation_context.cancelled);
+            REQUIRE(!ebpf_async_cancel(&completion_context));
+        } else {
+            REQUIRE(ebpf_async_cancel(&completion_context));
+            REQUIRE(completion_context.result == EBPF_PENDING);
+            REQUIRE(cancellation_context.cancelled);
+            REQUIRE(!ebpf_async_complete(&completion_context, EBPF_SUCCESS));
```

Comment:
`EBPF_CANCELLED`

Also a prescriptive guidance for handlers on how to handle cancel calls and that handlers must complete with CANCEL error code for canceled operations.